### PR TITLE
eager namespace refresher POC

### PR DESCRIPTION
**Pull request by xwduan**
**What changed?**
Add eater namespace refresher POC.

**Why?**
Handle the situation where the namespace failover version in current cluster is smaller than the History event version from the source cluster.

**How did you test it?**
N/A

**Potential risks**
N/A

**Is hotfix candidate?**
No

### CODE REVIEW

File: common/namespace/registry.go [hunk 1]
```diff
@@ -138,6 +138,7 @@ type (
 		GetNamespaceID(name Name) (ID, error)
 		GetNamespaceName(id ID) (Name, error)
 		GetCacheSize() (sizeOfCacheByName int64, sizeOfCacheByID int64)
+		RefreshNamespace(id ID)
```
Comment (wxing1292) [inline common/namespace/registry.go:141]: always return error

File: service/history/replication/eager_namespace_refresher.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+func (e *eagerNamespaceRefresherImpl) UpdateNamespaceFailoverVersion(namespaceId namespace.ID, targetFailoverVersion int64) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	ns, err := e.namespaceRegistry.GetNamespaceByID(namespaceId)
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NamespaceNotFound:
+		return nil
```
Comment (wxing1292) [inline service/history/replication/eager_namespace_refresher.go:34]: plz add a TODO here

File: service/history/replication/eager_namespace_refresher.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		Namespace:           resp.Namespace,
+		NotificationVersion: metadata.NotificationVersion,
+		IsGlobalNamespace:   resp.IsGlobalNamespace,
+	}
+
+	request.Namespace.FailoverVersion = targetFailoverVersion
+	request.Namespace.FailoverNotificationVersion = metadata.NotificationVersion
+
+	//Question: is it ok to only update failover version WITHOUT updating FailoverHistory?
+	//request.Namespace.ReplicationConfig.FailoverHistory = ??
```
Comment (wxing1292) [inline service/history/replication/eager_namespace_refresher.go:75]: @yux0

File: common/namespace/registry.go (hunk 1 above)
Comment (wxing1292) [reply common/namespace/registry.go:141]: let us try to not touch ns cache for now

File: service/history/replication/eager_namespace_refresher.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+	if err := e.metadataManager.UpdateNamespace(ctx, request); err != nil {
+		e.logger.Info("Failed to update namespace", tag.Error(err))
+		return err
+	}
+	e.namespaceRegistry.RefreshNamespace(ns.ID())
```
Comment (wxing1292) [inline service/history/replication/eager_namespace_refresher.go:81]: let us try not to update ns cache for now.

other part of the logic looks good, plz add unit tests

**Review by wxing1292 — APPROVED** [review]

**Review by yux0 — APPROVED** [review]

**Pull request merged**